File: src/test/java/com/projectswg/holocore/services/gameplay/combat/command/TestCombatCommandBuffNpcs.java
package com.projectswg.holocore.services.gameplay.combat.command;

import com.projectswg.holocore.resources.support.data.server_info.loader.ServerData;
import com.projectswg.holocore.resources.support.global.commands.CombatCommand;
import com.projectswg.holocore.resources.support.global.commands.Command;
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject;
import com.projectswg.holocore.resources.support.objects.swg.tangible.OptionFlag;
import com.projectswg.holocore.resources.support.objects.swg.weapon.WeaponObject;
import com.projectswg.holocore.services.gameplay.combat.buffs.BuffService;
import com.projectswg.holocore.test.resources.GenericCreatureObject;
import com.projectswg.holocore.test.runners.TestRunnerSynchronousIntents;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import static org.junit.Assert.assertEquals;

@RunWith(Parameterized.class)
public class TestCombatCommandBuffNpcs extends TestRunnerSynchronousIntents {
	
	private CreatureObject source;
	private CreatureObject target;
	
	@Parameterized.Parameter
	public Input input;
	
	@Parameterized.Parameters(name = "{0}")
	public static Collection<Input> input() {
		long group1 = 1234;
		long group2 = 4321;
		
		return Arrays.asList(
				new Input("Ungrouped player should not be able to buff ungrouped NPC", 0, 0, false,false),
				new Input("Grouped player should not be able to buff ungrouped NPC", group1, 0, false,false),
				new Input("Ungrouped player should not be able to buff grouped NPC", 0, group2, false,false),
				new Input("Grouped player should not be able to buff NPC in a different group", group1, group2, false,false),
				new Input("Grouped player should be able to buff NPC in the same group", group1, group1, false,true),
				new Input("Player should not be able to buff invulnerable NPC", 0, 0, true, false)
		);
	}
	
	@Before
	public void setup() {
		source = new GenericCreatureObject(1, "Player", true);
		source.setFaction(ServerData.INSTANCE.getFactions().getFaction("neutral"));
		source.setGroupId(input.getPlayerGroupId());
		
		// Create a weapon for the source creature. Is used during buff process and must not be null.
		WeaponObject weapon = new WeaponObject(2);
		weapon.setArrangement(List.of(List.of("hold_r")));
		weapon.systemMove(source);
		source.setEquippedWeapon(weapon);
		
		target = new GenericCreatureObject(3, "NPC", false);
		target.setFaction(ServerData.INSTANCE.getFactions().getFaction("neutral"));
		target.setGroupId(input.getNpcGroupId());
		
		if (input.isNpcInvulnerable()) {
			target.setOptionFlags(OptionFlag.INVULNERABLE);
		}
		
		registerService(new BuffService());
	}
	
	@Test
	public void testReceiveBuff() {
		String targetBuffName = "hemorrhage";	// Important that the buff actually exists
		Command command = Command.builder()
				.withName(targetBuffName)
				.build();
		
		CombatCommand combatCommand = CombatCommand.builder(command)
				.withBuffNameSelf("")
				.withBuffNameTarget(targetBuffName)
				.withDefaultAnimation(new String[]{""})
				.build();
		
		CombatCommandBuff.INSTANCE.handle(source, target, combatCommand, "");
		
		waitForIntents();	// Let's give the BuffService a chance to process the BuffIntent
		
		boolean expected = input.isExpected();
		String caseName = input.getCaseName();
		
		assertEquals(caseName, expected, target.hasBuff(targetBuffName));
	}
	
	private static class Input {
		private final String caseName;
		private final long playerGroupId;	// Group ID of the player
		private final long npcGroupId;	// Group ID of the NPC
		private final boolean npcInvulnerable;
		private final boolean expected;
		
		public Input(String caseName, long playerGroupId, long npcGroupId, boolean npcInvulnerable, boolean expected) {
			this.caseName = caseName;
			this.playerGroupId = playerGroupId;
			this.npcGroupId = npcGroupId;
			this.npcInvulnerable = npcInvulnerable;
			this.expected = expected;
		}
		
		public String getCaseName() {
			return caseName;
		}
		
		public long getPlayerGroupId() {
			return playerGroupId;
		}
		
		public long getNpcGroupId() {
			return npcGroupId;
		}
		
		public boolean isNpcInvulnerable() {
			return npcInvulnerable;
		}
		
		public boolean isExpected() {
			return expected;
		}
		
		@Override
		public String toString() {
			return caseName;
		}
	}
}

File: src/test/java/com/projectswg/holocore/services/gameplay/combat/command/TestCombatCommandBuffNpcs.kt
/***********************************************************************************
 * Copyright (c) 2024 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.services.gameplay.combat.command

import com.projectswg.holocore.resources.support.data.server_info.loader.ServerData.factions
import com.projectswg.holocore.resources.support.global.commands.CombatCommand
import com.projectswg.holocore.resources.support.global.commands.Command
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject
import com.projectswg.holocore.resources.support.objects.swg.tangible.OptionFlag
import com.projectswg.holocore.resources.support.objects.swg.weapon.WeaponObject
import com.projectswg.holocore.services.gameplay.combat.buffs.BuffService
import com.projectswg.holocore.test.resources.GenericCreatureObject
import com.projectswg.holocore.test.runners.TestRunnerSynchronousIntents
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import kotlin.collections.Collection
import kotlin.collections.listOf

class TestCombatCommandBuffNpcs : TestRunnerSynchronousIntents() {
	private var source: CreatureObject? = null
	private var target: CreatureObject? = null

	fun setup(input: Input) {
		// Create a weapon for the source creature. Is used during buff process and must not be null.
		val weapon = WeaponObject(2)
		weapon.arrangement = listOf(listOf("hold_r"))

		source = GenericCreatureObject(1, "Player", true).run {
			faction = factions.getFaction("neutral")
			groupId = input.playerGroupId

			equippedWeapon = weapon
			weapon.systemMove(this)
			return@run this
		}

		target = GenericCreatureObject(3, "NPC", false).run {
			faction = factions.getFaction("neutral")
			groupId = input.npcGroupId

			if (input.isNpcInvulnerable) setOptionFlags(OptionFlag.INVULNERABLE)
			return@run this
		}

		registerService(BuffService())
	}

	@ParameterizedTest
	@MethodSource("input")
	fun testReceiveBuff(input: Input) {
		setup(input)
		val commandName = "hemorrhage"
		val targetBuffName = "hemorrhage" // Important that the buff actually exists
		val command = Command.builder().withName(commandName).build()

		// @formatter:off
		val combatCommand = CombatCommand.builder()
			.withName(commandName)
			.withBuffNameSelf("")
			.withBuffNameTarget(targetBuffName)
			.withDefaultAnimation(arrayOf(""))
			.build()
		// @formatter:on

		CombatCommandBuff.INSTANCE.handle(source!!, target, command, combatCommand, "")

		waitForIntents() // Let's give the BuffService a chance to process the BuffIntent

		val expected: Boolean = input.isExpected
		val caseName: String = input.caseName

		Assertions.assertEquals(expected, target!!.hasBuff(targetBuffName), caseName)
	}

	class Input(val caseName: String, // Group ID of the player
		val playerGroupId: Long, // Group ID of the NPC
		val npcGroupId: Long, val isNpcInvulnerable: Boolean, val isExpected: Boolean) {
		override fun toString(): String {
			return caseName
		}
	}

	companion object {
		@JvmStatic
		fun input(): Collection<Input> {
			val group1: Long = 1234
			val group2: Long = 4321

			// @formatter:off
			return listOf(
				Input(
					caseName = "Ungrouped player should not be able to buff ungrouped NPC",
					playerGroupId = 0,
					npcGroupId = 0,
					isNpcInvulnerable = false,
					isExpected = false
				),
				Input(
					caseName = "Grouped player should not be able to buff ungrouped NPC",
					playerGroupId = group1,
					npcGroupId = 0,
					isNpcInvulnerable = false,
					isExpected = false
				),
				Input(
					caseName = "Ungrouped player should not be able to buff grouped NPC",
					playerGroupId = 0,
					npcGroupId = group2,
					isNpcInvulnerable = false,
					isExpected = false
				),
				Input(
					caseName = "Grouped player should not be able to buff NPC in a different group",
					playerGroupId = group1,
					npcGroupId = group2,
					isNpcInvulnerable = false,
					isExpected = false
				),
				Input(
					caseName = "Grouped player should be able to buff NPC in the same group",
					playerGroupId = group1,
					npcGroupId = group1,
					isNpcInvulnerable = false,
					isExpected = true
				),
				Input(
					caseName = "Player should not be able to buff invulnerable NPC",
					playerGroupId = 0,
					npcGroupId = 0,
					isNpcInvulnerable = true,
					isExpected = false
				)
			)
			// @formatter:on
		}
	}
}
